qiskit/providers/backend.py

```diff
@@ -229,3 +240,357 @@ class can handle either situation.
+    that the configuration attribute no longer exists instead attributes
+    exposing equivalent required immutable properties of the backend device
+    are added. For example ``backend.configuration().n_qubits`` is accessible
+    from ``backend.num_qubits`` now.
+
+    The ``options`` attribute of the backend is used to contain the dynamic
+    user configurable options of the backend. It should be used more for
+    runtime options that configure how the backend is used. For example,
+    something like a ``shots`` field for a backend that runs experiments which
+    would contain an int for how many shots to execute.
+
+    If migrating a provider from :class:`~qiskit.providers.BackendV1` or
+    :class:`~qiskit.providers.BaseBackend` one thing to keep in mind is for
+    backwards compatibility you might need to add a configuration method that
+    will build a :class:`~qiskit.providers.models.BackendConfiguration` object
+    and :class:`~qiskit.providers.models.BackendProperties` from the attributes
+    defined in this class for backwards compatibility.
+    """
+
+    version = 2
+
+    def __init__(
+        self,
+        provider,
+        name: str = None,
+        description: str = None,
+        online_date: datetime.datetime = None,
+        backend_version: str = None,
+        **fields,
+    ):
+        """Initialize a BackendV2 based backend
+
+        Args:
+            provider (Provider): A backwards reference to the
+                :class:`~qiskit.transpiler.Provider` object that the backend
+                is from
+            name: An optional name for the backend
+            description: An optional description of the backend
+            online_date: An optional datetime the backend was brought online
+            backend_version: An optional backend version string. This differs
+                from the :attr:`~qiskit.providers.BackendV2.version` attribute
+                as :attr:`~qiskit.providers.BackendV2.version` is for the
+                abstract :attr:`~qiskit.providers.Backend` abstract interface
+                version of the object while ``backend_version`` is for
+                versioning the backend itself.
+
+        Raises:
+            AttributeError: If a field is specified that's outside the backend's
+                options
+        """
+
+        self._options = self._default_options()
+        self._provider = provider
+        self._coupling_map = None
+        if fields:
+            for field in fields:
+                if field not in self._options.data:
+                    raise AttributeError("Options field %s is not valid for this backend" % field)
+            self._options.update_config(**fields)
+        self._basis_gates_all = None
+        self.name = name
+        self.description = description
+        self.online_date = online_date
+        self.backend_version = backend_version
+
+    @property
+    def instructions(self) -> List[Tuple[Instruction, Tuple[int]]]:
```

Comment:
So I love this method that let's me see all the instructions available on the backend
```py
[
 (Instruction(name='u', num_qubits=1, num_clbits=0, params=[Parameter(theta), Parameter(phi), Parameter(lambda)]),
  (1,)),
 (Instruction(name='cx', num_qubits=2, num_clbits=0, params=[]), (0, 1)),
 (Instruction(name='cx', num_qubits=2, num_clbits=0, params=[]), (1, 0)),
 (Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]), (0,)),
 (Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]), (1,)),
 (Instruction(name='ecr', num_qubits=2, num_clbits=0, params=[]), (1, 0))]
```

However I don't know how to read the error/duration/calibration for one of these instructions that interests me. It would have been nice to be able to do something like:
`target.instructions(3).calibration`, similar to the syntax we discussed for qubits `backend.qubits(3).t1`.

Assuming a backend has a finite number of instructions and qubits available, I think it would be nice to grab them by some index and look up their properties.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/debae293897c7b8fc576cd455635c798f5439516